include/oneapi/dpl/pstl/execution_impl.h

```diff
@@ -105,25 +78,163 @@ template <typename _ExecutionPolicy>
+struct __parallel_tag
+{
+    using __is_vector = _IsVector;
+    // backend tag can be change depending on
+    // TBB availability in the environment
+    using __backend_tag = __par_backend_tag;
+};
+
+struct __parallel_forward_tag
+{
+    using __is_vector = ::std::false_type;
+    // backend tag can be change depending on
+    // TBB availability in the environment
+    using __backend_tag = __par_backend_tag;
+};
+
+template <class _IsVector, class... _IteratorTypes>
+using __tag_type = ::std::conditional_t<
+    __internal::__is_random_access_iterator_v<_IteratorTypes...>, __parallel_tag<_IsVector>,
+    ::std::conditional_t<__is_forward_iterator_v<_IteratorTypes...>, __parallel_forward_tag, __serial_tag<_IsVector>>>;
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::sequenced_policy>,
+    __serial_tag<std::false_type>>
+__select_backend()
 {
-    return __internal::__lazy_and(__exec.__allow_parallel(),
-                                  __internal::__is_random_access_iterator_t<_IteratorTypes...>());
+    return {};
 }
 
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::unsequenced_policy>,
+    __serial_tag<__internal::__is_random_access_iterator<_IteratorTypes...>>>
+__select_backend()
+{
+    return {};
+}
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::parallel_policy>,
+    __tag_type<std::false_type, _IteratorTypes...>>
+__select_backend()
+{
+    return {};
+}
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::parallel_unsequenced_policy>,
+    __tag_type<__internal::__is_random_access_iterator<_IteratorTypes...>, _IteratorTypes...>>
+__select_backend()
+{
+    return {};
+}
+
+//----------------------------------------------------------
+// __is_backend_tag, __is_backend_tag_v
+//----------------------------------------------------------
+
+template <class _Tag>
+struct __is_backend_tag : ::std::false_type
+{
+};
+
+template <class _IsVector>
+struct __is_backend_tag<__serial_tag<_IsVector>> : ::std::true_type
+{
+};
+
+template <class _IsVector>
+struct __is_backend_tag<__parallel_tag<_IsVector>> : ::std::true_type
+{
+};
+
+template <>
+struct __is_backend_tag<__parallel_forward_tag> : ::std::true_type
+{
+};
+
+template <class _Tag>
+inline constexpr bool __is_backend_tag_v = __is_backend_tag<::std::decay_t<_Tag>>::value;
+
+//----------------------------------------------------------
+// __is_backend_tag_serial, __is_backend_tag_serial_v
+//----------------------------------------------------------
+
+template <class _Tag>
```

Comment:
1) In my understanding we need just lines 208-216. And we dont need __is_backend_tag_serial_v because that traits is used just in assert.. I think it doesnt make sense to introduce additional things like `__is_backend_tag_serial_v`
2) Other lines 182-207 and 217 - 237 should  be removed.

Reply:
As we discussed offline `__is_backend_tag_parallel_forward` is also useful.
But, the rest traits defined "nearby" are not used, it seems.

Reply:
Fixed